### DISCUSSION TITLE
Shadow and tornettools don't seem to handle packet loss well.

### DISCUSSION THREAD

**Discussion by higher-order** [shadow/shadow#3325] [Q&A] [answered]
I'm simulating networks generated by [tornettools](https://github.com/shadow/tornettools), and experimenting with changing the `packet_loss` parameters on the edges in the network graph `atlas_v201801.shadow_v2.gml`. The graph has `packet_loss 0.0` for all edges. I'm doing a sanity check to see if setting non-zero loss rates will indeed negatively affect the performance.

The issue was first noticed when after changing all `packet_loss` to 0.011, running the otherwise same 1% Tor net for a simulation time of 60 minutes produced completely identical tornettools plots (using the `tornettools plot` command). Only that the simulation took 26 hours to finish, instead of just 3 hours if all `packet_loss` are 0.0.

I don't have the plots of that experiment now, but do have plots for simulations on the same 1% Tor net with `packet_loss` set to:
* 0.0 for all edges (no loss);
* 0.011 for a random selection of 10% of all edges;
* 0.0002 for a random selection of 5% of all edges.
  * 0.0002 seems close to a realistic loss, which is collected from a realistic network graph I have.

I ran the simulations for 900 seconds each. But the plots are identical again. Except that the simulations with loss are slower.

The only other discussion I found related to "loss" is [this](https://github.com/shadow/shadow/discussions/576). It is 10+ years ago and seems to deal with TCP autotuning issues.

I'd like to ask if I'm configuring loss in a wrong way, or is it more of a known issue in Shadow?
Thank you!

Edit: I've also run `iperf2` server and client on two of the virtual hosts in the network with 0.011 loss on all edges, and confirmed that it showed a much lower bandwidth (about 200x lower). But the plots are the same.

Edit: ~~I'll upload the 3 plots in just a short while.~~ The PDFs are uploaded.
[tornet.plot.pages.pdf](https://github.com/shadow/shadow/files/14937092/tornet.plot.pages.pdf)
[tornet.plot.pages.pdf](https://github.com/shadow/shadow/files/14937096/tornet.plot.pages.pdf)
[tornet.plot.pages.pdf](https://github.com/shadow/shadow/files/14937100/tornet.plot.pages.pdf)

**Comment by jtracey** [comment]
Note that the lack of packet loss is just packet loss on the links, there will be packet loss from congestion even when all the links have 0 packet loss. For why you're not seeing any differences, I suspect that might have to do with this being on a 1% network. You can see the [Once is Never Enough](https://cypherpunks.ca/~iang/pubs/neverenough-usenix21.pdf) paper for why we set packet loss to 0, and an explanation for why 1% networks can behave very strangely (3.1 and 6.3, respectively).

**Comment by higher-order** [comment, marked as answer]
Turns out tornettools wasn't plotting the lines for the simulation... It only plotted the public Tor lines. That's why the plots looked identical.
